src/main/java/picard/sam/util/PrimaryAlignmentKey.java

```diff
@@ -0,0 +1,52 @@
+package picard.sam.util;
+
+import htsjdk.samtools.SAMRecord;
+
+/**
+ * It is useful to define a key such that the key will occur at most once among the primary alignments in a given file
+ * (assuming the file is valid). The read name + pairing status should be sufficient for this.
+ */
+public final class PrimaryAlignmentKey implements Comparable<PrimaryAlignmentKey> {
+
+    private enum PairStatus {UNPAIRED, FIRST, SECOND} // note the order here; it should correspond to that of SAMRecordQueryNameComparator!
+
+    private final PairStatus pairStatus;
+    private final String readName;
+
+    public PrimaryAlignmentKey(final SAMRecord rec) {
+        if (rec.isSecondaryOrSupplementary()) {
+            throw new IllegalArgumentException("PrimaryAligmentKey cannot be a secondary or suplementary alignment");
+        }
+        this.pairStatus = rec.getReadPairedFlag() ?
+                (rec.getSecondOfPairFlag() ? PairStatus.SECOND : PairStatus.FIRST) :
+                PairStatus.UNPAIRED;
+        this.readName = rec.getReadName();
+    }
+
+    @Override
+    public int hashCode() {
+        return 31 * this.readName.hashCode() + this.pairStatus.hashCode();
```

Comment:
I believe that this is the same as `Objects.hash(readName, pairStatus);`

Reply:
Done in previous round.